### PR TITLE
#89/feat/adding-new-tags

### LINKED ISSUE #89: Create functionality for adding tags to system

### PR DESCRIPTION
**Pull request by piamanns**
Closes #89 

### In this pull request: 
- A separate page for adding new tags to the system
- The form on the page won't let the user add an empty tag or a tag that is over 50 characters long
- Trying to add a duplicate tag to the database (regardless of casing) results in an error message
- The page uses the app-wide notification system snackbar for displaying notifications and error messages

### CODE REVIEW

File: src/components/TagForm/index.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    register,
+    formState: { errors, isSubmitting },
+    handleSubmit,
+    reset,
+  } = useForm<TagSchemaType>({
+    resolver: zodResolver(tagSchema),
+  });
+
+  const submitForm = async (data: TagSchemaType) => {
+    const responseJson = await post('/api/tag', data);
```
Comment (anketola) [inline src/components/TagForm/index.tsx:26]:
```suggestion
    data.name = data.name.replace(/\s{2,}/g, ' ').trim();
    const responseJson = await post('/api/tag', data);
```
You might want something like that to prevent duplicates better. The regex prevents adding for example "Agile Methods" and "Agile  Methods" (two spaces between there). Then it trims trailing and leading spaces to prevent stuff like " Docker", "Docker  " etc.

File: src/lib/zod/tags.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import z from 'zod';
+
+export const tagSchema = z.object({
+  name: z
+    .string()
+    .min(1, 'A tag name is required')
+    .max(50, 'The maximum length for a tag is 50 characters'),
+});
```
Comment (anketola) [inline src/lib/zod/tags.ts:8]:
```suggestion
 name: z
    .string()
    .refine(
      (value) => value.trim() !== '',
      { message: 'A tag name is required' }
    )
    .refine(
      (value) => value.trim().length <= 50,
      { message: 'The maximum length for a tag is 50 characters' }
    )
    .refine(
      (value) => !/\s{2,}/.test(value),
      { message: 'Consecutive spaces are not allowed' }
    ),
});
```

This should, at least to my knowledge, work on Zod.  Should do the same as the regex on TagForm component.

**Review by anketola — CHANGES_REQUESTED** [review]
Good job, all the functionality is there.

You might want to check the suggested changes to prevent bit more the duplicates. Another thing is you could have a look if you want to add some component level tests, currently there are none.

File: src/components/TagForm/index.tsx (hunk 1 above)
Comment (piamanns) [reply src/components/TagForm/index.tsx:27]: Very good addition! Extra spaces can cause a lot of trouble; a managed to miss that.

File: src/lib/zod/tags.ts (hunk 2 above)
Comment (piamanns) [reply src/lib/zod/tags.ts:8]: Yes, definitely worth implementing!

File: src/components/TagForm/index.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    notify(responseJson);
+    reset();
+    router.refresh();
+  };
+
+  return (
+    <>
+      <Box sx={{ mt: 1, mb: 4 }}>
+        <form onSubmit={handleSubmit(submitForm)}>
+          <TextField label="Tag name" {...register('name')}></TextField>
```
Comment (anketola) [inline src/components/TagForm/index.tsx:37]:
```suggestion
          <TextField label="Tag name" {...register('name')} inputProps={{ maxLength: 51 }}></TextField>
```

If you want, you can just have in frontend the user stop writing once the message error has appeared with the inputProps maxlenght attribute. Right now the use can keep typing despite the lenght error message. If you choose to do that, not having it as a magic number would be better though. Up to you.

**Review by anketola — APPROVED** [review]
Looks peachy.

Zod 3.22.2 has redos vulnerability in email validation. We don't have anything related to that (although I tested and it was fun breaking stuff), but might have later with new features coming. It's fixed in 3.22.3+

**Review by mmoila — APPROVED** [review]
Looking good! 👍

File: src/components/TagForm/index.tsx (hunk 3 above)
Comment (piamanns) [reply src/components/TagForm/index.tsx]: Good idea, I implemented this (and got rid of the magic number)
